feat: Commands mode supports arguments

klesh opened the pull request:
# Description

Implementation of #11897 to support CLI tools like `lf`.

## Use case

This use case demonstrates how Nushell can seamlessly integrate with CLI tools like lf that support user-defined custom scripts.

`lf` Custom Script Execution with Nushell

`lf` allows defining custom commands that are executed using the given default shell. Here's how Nushell can be used within an `lf` script:

### Example: Creating Executable Nushell Scripts

Imagine an `lf` command named `mk-nu-scripts` that creates executable Nushell scripts:
```
cmd mk-nu-scripts &{{
  for x in $args {
    "#!/usr/bin/env nu" | save $x
    chmod +x $x
  }
}}
```
This command iterates over provided arguments (`$args`) and generates new files with the shebang line `#!/usr/bin/env nu`, effectively turning them into executable Nushell scripts.

### Usage within lf:

Simply execute the command with desired filenames as arguments:
```
:mk-nu-scripts foo bar
```

### Behind the Scenes:

`lf` automatically separates arguments. The `--` symbol ensures any arguments following it are passed to the Nushell script itself (specified by `-c`) and not interpreted by the shell.

### Benefits:

- Enhanced Workflow: Automate repetitive tasks within lf using powerful Nushell scripting.
- Extended Functionality: Leverage Nushell's capabilities to create custom lf features.
- Improved Developer Experience: Utilize existing Nushell knowledge within lf.

### Nushell Integration Potential:

This example showcases the potential of Nushell's integration with lf and similar CLI tools. Imagine building complex workflows and extending functionality seamlessly across different tools.

# User-Facing Changes
<!-- List of all changes that impact the user experience here. This helps us keep track of breaking changes. -->
Be able to use `nu` shell as the **default shell**

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use std testing; testing run-tests --path crates/nu-std"` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->

# After Submitting
<!-- If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date. -->

fdncred commented:
What exactly is this trying to add? I'm not following. Once can already use nushell as a default shell. I also have no idea what `lf` is.

39555 commented:
Hi! It seems you attached the wrong link to an unrelated dependabot issue? What kind of support have you implemented? Any examples with lf? Any tests?

Context: https://github.com/gokcehan/lf

39555 commented:
A okay, I tested it myself. `lf` sends all commands with a double dash at the end.
Double dash "--" is not supported yet.

#7898 
#10939 
#11435
#8323

```
Error: nu::parser::unknown_flag

  × The `nu` command doesn't have flag ``.
   ╭─[source:2:8]
 1 │ nu -c "IFS='
 2 │ '; ls" --
   ·        ─┬
   ·         ╰── unknown flag
   ╰────
```

klesh commented:
Sorry, folk, I put the wrong issue number, it should be https://github.com/nushell/nushell/issues/11897

klesh commented:
@39555 Thanks for the clarification. I think `lf` is just one of the CLI tools that need to invoke the default shell with `commands`  and pass args to the `commands` instead of the  `shell`.

@fdncred Sorry for the confusion. 😂 `lf` is a CUI file management tool.

klesh commented:
Hi, folks. Is there anything I can do to move this forward? I really like `nushell` and it would be nice if I could use it as the default shell for other programs like `lf` so anyone could write scripts in `nushell`  that work cross-platform-ly.

For example, [lf](https://github.com/gokcehan/lf) is a CLI file manager. One may write custom commands in it and they would be executed in the default shell. The default shell on Linux may be `sh`, and `cmd` on Windows, so it is fine for people who use only any one of them. However, I'm using both, it would be nice if I could set the default shell to `nu` and write scripts for multiple platforms, and I believe it is a nice thing to have.

devyn commented:
I like the idea of this, but I wonder if just for the sake of correctness it might be good to use the existing `def main` syntax for scripts, even if it's a bit longer. It would allow you to get type validation/conversion if that's useful for you, e.g.

```nushell
nu -c 'def main [foo: int, bar: int] { print ($foo + $bar) }' 1 2
```

This would return a much better error if the arguments aren't ints, or if there aren't enough arguments, than just accessing `$args`

I know `--commands` isn't really meant for *serious* scripting, and so this might be more of an annoyance - but I'm just putting it out there.

devyn commented:
I suppose if I understand your use case correctly, you probably really would rather not have to write `def main`. But in that case I don't think we should try to handle types, and just always treat them all as strings.

`--wrapped` or rest args not necessary in that case - can just generate a list in Rust and send that as the argument.

kubouch commented:
It is quite hard to understand what the PR is trying to achieve. Having some examples of the desired functionality and more general description would help. I have never heard about lf.

That said, I don't believe the PR is doing an equivalent of `sh -c "echo $@"

kubouch commented:
It is quite hard to understand what the PR is trying to achieve. Having some examples of the desired functionality and more general description would help. I have never heard about lf.

That said, I don't believe the PR is doing an equivalent of `sh -c "echo $@" -- foo bar` because we don't have an equivalent of `$@`. It seems to simply append the args after `--` to whatever was inside the `-c` flag ( the `def --wrapped` seems unnecessary to me). As such, there doesn't seem to be much of a difference between `nu -c "echo" -- foo bar` and `nu -c "echo foo bar"`, so adding support for `--` seems unnecessary to me because it further complicates the `nu` parsing rules and making them more different from our language parsing rules than they already are. There was a discussion about adding `--` to the language but that's for a different purpose.

klesh commented:
@kubouch Hi, thanks for the comment, I just updated the description and added a Use Case section, would you like to take a look? Thanks in advance.

@devyn Will `a list in Rust` be simpler? or `--wrapped` or rest args are problematic?

klesh commented:
@fdncred I just updated the description, would you like to take a look as well? Thanks.

fdncred commented:
if i understand your use case, your lf script is creating two scripts, foo and bar and all they have in them is the shebang. So, kind of just initializing scripts with boiler plate, right?

what other cli tools would benefit from nushell having a -- command line parameter like this?

devyn commented:
The gist of this PR is that it seems that some tools will allow you to run your shell with whatever command you want and also pass additional args to your shell to avoid having to know how to escape them properly.

I think this is a nice thing to be able to do because you don't really want to have to try to escape strings just to get them into nushell. We support this for script files, so why not for `--commands`?

The main thing I worry about is that if we're doing this, it should actually be better and prevent injection by directly constructing the values.

@klesh I definitely think you should try to compile a command from it and then evaluate it with a list in rust. There's not really any reason to take the extra step of using `--wrapped` in that case.

klesh commented:
@fdncred Yes, correct. However, it is NOT a real use case, I just cooked it up according to what I learned from https://github.com/gokcehan/lf/pull/1606 , the point is that scripts (`commands`) might need arbitrary `arguments`. It might be `--` or other symbols to indicate the `args` are meant for `commands` not the `shell` itself.

in the following example, it is hard to tell who the `foo` and `bar` are meant for, the `commands` or the `nu` binary?
`nu -c 'echo $args' foo bar`

many shells would use `--` as an indicator to distinguish who is the rest of the args meant for.

`nu -c 'echo $args' foo bar` => `foo` and `bar` are meant for the `nu`
`nu -c 'echo $args' -- foo bar` => `foo` and `bar` are meant for the `commands` script
`nu -c 'echo $args' foo -- bar` => `foo` is meant for the `nu` while `bar` is meant for the `commands` script


`kubectl` uses this convention as well https://kubernetes.io/docs/tasks/debug/debug-application/get-shell-running-container/



@devyn I think I understand your point, to evaluate commands with a strings list to prevent script injection, and I agree. Would you mind showing me some existing code as an example since I'm not very familiar with the codebase yet.😀 Thanks in advanced.

devyn commented:
@klesh sure, happy to try to cook something up tomorrow.

kubouch commented:
I just realized you can use `$args` which makes it a Nushell equivalent of `$@`. This makes it actually useful. It's still extremely hacky, but also quite cool.

One issue with it is that we had a proposal for adding `--` to custom commands that would mean “parse everything after `--` as strings and collect to `...rest`”. That would mean that `--` would behave differently on the `nu` executable and in our custom command call syntax.

We should also weigh in the corner cases: How would `--` behave without the `-c` flag? How would it interact with scripts, e.g., `nu script.nu -- arg1 arg2`?

And also, do we _need_ this feature for 1.0? I feel like getting more POSIX support is always nice, _if it doesn't clash with Nushell's own way_, so I think the corner cases should be analyzed carefully.

devyn commented:
Hey @klesh. In looking into this more, I discovered that there's a function we already trust to escape args for scripts, so it seems like we can use the same one. I made a suggestion in the review for how to do this.

klesh commented:
@devyn I don't see the review comment, did you submit the review?

klesh commented:
@kubouch Thanks for the information, where can I find this `...rest` proposal?

devyn commented:
@klesh My bad, see review :point_up:

klesh commented on crates/nu-cli/src/eval_cmds.rs:
```diff
@@ -33,7 +34,17 @@ pub fn evaluate_commands(
+            commands = format!(
+                "def --wrapped wrapped_commands [...args] {{ {} }}; wrapped_commands {}",
+                commands,
+                args_to_commands.join(" "),
```
Agree, I was planning to use the escaped single-quoted strings to address the problem. Sadly, it was not possible, https://github.com/nushell/nushell/issues/12343 .😂

klesh replied:
Done

klesh commented on src/command.rs:
```diff
@@ -7,11 +7,13 @@ use nu_protocol::{
 use nu_utils::stdout_write_all_and_flush;
 
-pub(crate) fn gather_commandline_args() -> (Vec<String>, String, Vec<String>) {
+pub(crate) fn gather_commandline_args() -> (Vec<String>, String, Vec<String>, Vec<String>) {
```
I agree. Should it be done in this PR or another one?

devyn replied:
It's fine to do it in this PR. You're adding something anyway, and this is entirely local to this file

klesh commented on crates/nu-cli/src/eval_cmds.rs:
```diff
@@ -33,7 +34,17 @@ pub fn evaluate_commands(
-        let output = parse(&mut working_set, None, commands.item.as_bytes(), false);
+        let mut commands = commands.item.clone();
+        if !args_to_commands.is_empty() {
+            // Should args_to_commands be escaped in case some of them have spaces??
+            // It would be easier if https://github.com/nushell/nushell/issues/12343 could be addressed first.
+            commands = format!(
+                "def --wrapped wrapped_commands [...args] {{ {} }}; wrapped_commands {}",
+                commands,
+                args_to_commands.join(" "),
+            )
```
Thanks, very helpful.

klesh replied:
Done

klesh commented:
Hi, @devyn . Thanks for the help. I've just updated the code, would you mind taking another look. Thanks in advanced.

devyn commented on src/main.rs:
```diff
@@ -152,8 +152,11 @@ fn main() -> Result<()> {
-
-    let (args_to_nushell, script_name, args_to_script) = gather_commandline_args();
+    let commandline_args = gather_commandline_args();
+    let args_to_nushell = commandline_args.args_to_nushell;
+    let script_name = commandline_args.script_name;
+    let args_to_script = commandline_args.args_to_script;
+    let args_to_commands = commandline_args.args_to_commands;
```
You can destructure this using a pattern on the left-hand side of the `let`

klesh replied:
Done

devyn reviewed (commented):
Looks better to me for sure. I'll give it a run later to see what it's like to use.

klesh commented:
> I just realized you can use `$args` which makes it a Nushell equivalent of `$@`. This makes it useful. It's still extremely hacky, but also quite cool.
> 
> One issue with it is that we had a proposal for adding `--` to custom commands that would mean “parse everything after `--` as strings and collect to `...rest`”. That would mean that `--` would behave differently on the `nu` executable and in our custom command call syntax.
> 

I kind of feel that it is OK, take the `cargo` itself as an example. I use the following command to test out if the patch is working properly:
```
 cargo run -- -c "echo $args" -- foo bar "hello world"
```
The first `--` tells `cargo` to pass the rest of the arguments to the compiled binary while the second one tells the `nu` binary to pass the rest of the arguments to the `commands`. I think It is OK because there are no ambiguous for people to understand it correctly. This is under the 2 binary executable context.

In the context of the `executable` and `commands`, they are working in different mechanisms with different designs, I think they could interpreted the `--` differently?


> We should also weigh in the corner cases: How would `--` behave without the `-c` flag? How would it interact with scripts, e.g., `nu script.nu -- arg1 arg2`?
> 

Make sense, I think it should be up to the `script.nu` to decide based on what it tries to achieve, not sure. Maybe it could be addressed in another thread?

> And also, do we _need_ this feature for 1.0? I feel like getting more POSIX support is always nice, _if it doesn't clash with Nushell's own way_, so I think the corner cases should be analyzed carefully.

I don't know, but it certainly would help me since I really need this.

devyn commented:
Hey @klesh, thanks for getting this into a good state. I think we still don't have consensus on exactly how this fits in with our design, but we'll probably talk about it at our meeting on Wednesday. I still think that some version of this would be useful, but we need to talk about how to design it in a way that it won't cause surprising behaviour.

klesh commented:
@devyn Sure, no problem. Thanks for your help.

klesh commented:
Hi, all. Any news?

devyn commented:
Ah, sorry, let me sync up with everyone else on this again. I'll let you know how things look after I gather enough eyes on it.

klesh commented:
@devyn I see, looking forward to it. Thanks.

devyn commented:
Okay, so basically as I understand it, we just want `--` behaviour here to be consistent with whatever we end up doing for internal commands. It sounds like `--` should mean that we stop parsing named arguments.

So I'm kind of thinking that this should still pass any other rest args to `$args` for `-c` even without `--`, and just not expect a filename at all. That might be a little complex though

Ideally, we can just replace this with the implementation in the parser whenever that gets added, so then `--` doesn't need to be handled specially here at all.

klesh commented:
> Okay, so basically as I understand it, we just want `--` behaviour here to be consistent with whatever we end up doing for internal commands. It sounds like `--` should mean that we stop parsing named arguments.

Basically yes, but I am not quite sure what  "internal commands" means. 😂

> 
> So I'm kind of thinking that this should still pass any other rest args to `$args` for `-c` even without `--`, and just not expect a filename at all. That might be a little complex though

Yes, I agree, especially when `-c` can be used in conjunction with `-l` or other args that the `nushell` binary has already supported or will support in the future. I think using `--` would be more reliable because it clearly states the intention.

> 
> Ideally, we can just replace this with the implementation in the parser whenever that gets added, so then `--` doesn't need to be handled specially here at all.

Which parser are you referring to? Can you share the link to the parser? Thanks.

devyn commented:
Internal commands are those defined within nushell (i.e. builtin, plugin, and custom commands), as opposed to externals, which run external processes to function. We want to support `--` for those at some point and there was concern that this wouldn't be consistent with how that would work

As for the parser, I'm just talking about the nu parser - the one in `crates/nu-parser`.

klesh commented:
Aah, I see. Thanks for the explanation.

I agree on maintaining consistency between internal commands and externals. However, enforcing a strict rule might be impractical due to their different execution contexts. nu is a binary executed by the operating system, while internal commands are defined and run within Nushell itself.

Could you share the specific logic you prefer? I would like to take a look at it and see what I can do.

devyn commented:
It's actually not about internals and externals (though we want to try to follow whatever is conventional to some degree), but more about making sure that `nu` is consistent with its own parsing for internal commands. The args to `nu` are actually implemented via the parser too anyway, but it would be surprising if it didn't work like Nushell commands usually do.

Of course, we can't guarantee that some random binary won't use `--` in an unusual way, and we won't try to. But at least for what we can control we want it to be unsurprising.

klesh commented:
OK, to address the problem,  I think we should take the `script` mode into consideration as well. When we type `nu -h` we would get
```
klesh@klesh-archpc ~> nu -h
The nushell language and shell.

Usage:
  > nu {flags} (script file) ...(script args) 
```

So, based on the above information and your comment. Here is what I have in mind, feel free to correct me if I am wrong:

- `--` should be optional, when omitted, all non-options (the ones without `-` or `--` prefixes) should be passed into `commands` or `script file` as `script args` while options should be sent to the `nu` binary.
- when `--` presented: all args (option and non-option)  before `--` are meant for `nu` binary, while the rest are meant for `commands` or `script`
- `commands` and `script` are mutually exclusive, the priority of `-c` is higher than `script` mode.


> The args to nu are actually implemented via the parser too anyway, but it would be surprising if it didn't work like Nushell commands usually do.

I can't relate the statement, I didn't see the args to nu going through the parser in the source code, they are processed in the `src/command.rs` if I understand it correctly:
![image](https://github.com/nushell/nushell/assets/61080/e7b9ec72-7f1a-41bc-91e0-26aa04479ae5)

Would you mind pointing me to the code? Thanks

klesh commented:
Hi, @devyn would you like to take a look and point me in some directions when you find the time? I believe the idea of the feature could benefit people and the nutshell community, the implementation may not be the best or even good enough, but I am willing to work on it, please help.

fdncred commented:
Where are we at with this? @devyn do you have time look into this again?

devyn commented:
Ah, sorry, I should have given an update. Last I checked in, we were talking about it, but I think we need a pattern that can be consistent with other commands - which we don't have yet, but planning for.

In general, I'm thinking the rule should be that anything after `--` goes into the rest parameter (with options unparsed), and `-c` causes there to be an implicit rest parameter named `$args`. So, these two would be basically the same:

```nushell
foo -- bar baz quux # proposed syntax
foo ...[bar baz quux] # works today
```

It may be better if we just make this change to the parser first. I'm thinking it also probably shouldn't affect `--wrapped` or external calls, because there are external commands that need the `--`. If we make that change, we can avoid having to do very much special here: `-c` just needs to take that rest parameter implicitly, as if it were

```nushell
def main [...args: string] {
  ...
}
```

WindSoilder commented:
Thanks for proposing the PR. I think it's too old and have conflicts.
I'm going to close it. If you think the change is still necessary, feel free to resolve the conflict and open a new PR.
